# HoMer_workflow

This repository contains the commands and scripts for running HoMer ([website](https://compbio.engr.uconn.edu/software/homer/), [paper](https://academic.oup.com/mbe/article/38/6/2639/6132264)) and RANGER-DTL ([website](https://compbio.engr.uconn.edu/software/RANGER-DTL/), [paper](https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty314/4983062?guestAccessKey=f95d29d3-0976-4c4d-b2ea-139dabd24bf8)). RANGER-DTL reconciles gene trees with species (or strain) trees to identify potential duplication, transfer, or loss events that occurred. HoMer builds upon this output, along with additional genome annotation and synteny information, to identify putative multi-gene transfer events.

**Please note that this repository is _not_ an official workflow provided by the tool developers**. This is simply the workflow I came up with for running HoMer, which required troubleshooting and custom scripts to produce the final results. My intention here is to provide a common repository that can be cited for multiple projects worked on by myself and colleagues, but this may be useful to others as well.

Likely the most important thing to keep in mind is that HoMer does not allow *any* non-alphanumeric characters (e.g., underscores) in any input file paths. This is true for both the file names themselves *and for the full file paths* (e.g., no underscores or periods can be present in any folder names that are part of file paths).
